src/stream/src/executor/sink.rs

```diff
@@ -133,22 +172,68 @@ impl SinkExecutor {
                         // thus mark the txn as started.
-                        if !in_transaction {
-                            sink.begin_epoch(epoch).await?;
-                            in_transaction = true;
-                        }
-
-                        if let Err(e) = sink.write_batch(chunk.clone()).await {
-                            sink.abort().await?;
-                            return Err(e.into());
-                        }
-                        empty_checkpoint_flag = false;
+                        log_writer.write_chunk(chunk.clone()).await?;
 
                         yield Message::Chunk(chunk);
                     }
                 }
                 Message::Barrier(barrier) => {
-                    if barrier.checkpoint {
+                    log_writer
+                        .flush_current_epoch(barrier.epoch.curr, barrier.checkpoint)
+                        .await?;
+                    if let Some(vnode_bitmap) = barrier.as_update_vnode_bitmap(actor_context.id) {
+                        log_writer.update_vnode_bitmap(vnode_bitmap);
+                    }
+                    yield Message::Barrier(barrier);
+                }
+            }
+        }
+    }
+
+    async fn execute_consume_log<R: LogReader>(
+        config: SinkConfig,
+        schema: Schema,
+        pk_indices: Vec<usize>,
+        connector_params: ConnectorParams,
+        sink_type: SinkType,
+        mut log_reader: R,
+        sink_commit_duration_metrics: Histogram,
+    ) -> StreamExecutorResult<Message> {
+        let mut sink = build_sink(config, schema, pk_indices, connector_params, sink_type).await?;
+
+        let mut epoch = log_reader.init().await?;
+
+        // the flag is required because kafka transaction requires at least one
+        // message, so we should abort the transaction if the flag is true.
+        let mut empty_checkpoint_flag = true;
+        let mut in_transaction = false;
+
+        loop {
+            let item: LogStoreReadItem = log_reader.next_item().await?;
+            match item {
+                LogStoreReadItem::StreamChunk(chunk) => {
+                    // NOTE: We start the txn here because a force-append-only sink might
+                    // receive a data chunk full of DELETE messages and then drop all of them.
+                    // At this point (instead of the point above when we receive the upstream
+                    // data chunk), we make sure that we do have data to send out, and we can
+                    // thus mark the txn as started.
+                    if !in_transaction {
+                        sink.begin_epoch(epoch).await?;
+                        in_transaction = true;
+                    }
+
+                    if let Err(e) = sink.write_batch(chunk.clone()).await {
+                        sink.abort().await?;
+                        return Err(e.into());
+                    }
```

Comment:
Not so related to the PR, but shall we handle the situation where the sink aborts after the implementation of kv log store, like retrying?